llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,251 @@
+/// level. Re-using it for complex CFG with arbitrary memory paths is definitely
+/// not recommended.
+Value *stripToMemorySource(Value *V) {
+  Value *MemoryAccess = V;
+  if (auto *LI = dyn_cast<LoadInst>(MemoryAccess)) {
+    Value *LoadSource = LI->getPointerOperand();
+    auto *Store = cast<StoreInst>(*llvm::find_if(
+        LoadSource->users(), [](User *U) { return isa<StoreInst>(U); }));
+    MemoryAccess = Store->getValueOperand();
+  }
+  return MemoryAccess->stripPointerCastsAndAliases();
+}
+
+void emitError(Function *PrintfInstance, CallInst *PrintfCall,
+               StringRef RecommendationToUser = "") {
+  std::string ErrorMsg =
+      std::string("experimental::printf requires format string to reside "
+                  "in constant "
+                  "address space. The compiler wasn't able to "
+                  "automatically convert "
+                  "your format string into constant address space when "
+                  "processing builtin ") +
+      PrintfInstance->getName().str() + " called in function " +
+      PrintfCall->getFunction()->getName().str() + ".\n" +
+      RecommendationToUser.str();
+  PrintfInstance->getContext().emitError(PrintfCall, ErrorMsg);
+}
+
+/// This routine goes over CallInst users of F, resetting the called function
+/// to CASPrintfFunc and generating/retracting constant addrspace format
+/// strings to use as operands of the mutated calls.
+size_t setFuncCallsOntoCASPrintf(Function *F, Function *CASPrintfFunc,
+                                 FunctionVecTy &FunctionsToDrop) {
+  size_t MutatedCallsCount = 0;
+  SmallVector<std::pair<CallInst *, Constant *>, 16> CallsToMutate;
+  for (User *U : F->users()) {
+    if (!isa<CallInst>(U))
+      continue;
+    auto *CI = cast<CallInst>(U);
+
+    // This key algorithm reaches the global string used as an argument to a
+    // __spirv_ocl_printf call. It then generates a constant AS copy of that
+    // global (or gets an existing one). For the return value, the call
+    // instruction is paired with its future constant addrspace string
+    // argument.
+    Value *Stripped = stripToMemorySource(CI->getArgOperand(0));
+    if (auto *Literal = dyn_cast<GlobalVariable>(Stripped))
+      CallsToMutate.emplace_back(CI, getCASLiteral(Literal));
+    else if (auto *Arg = dyn_cast<Argument>(Stripped)) {
+      // The global literal is passed to __spirv_ocl_printf via a wrapper
+      // function argument. We'll update the wrapper calls to use the builtin
+      // function directly instead.
+      Function *WrapperFunc = Arg->getParent();
+      std::string BadWrapperErrorMsg =
+          "Consider simplifying the code by "
+          "passing format strings directly into experimental::printf calls, "
+          "avoiding indirection via wrapper function arguments.";
+      if (!WrapperFunc->getName().contains("6oneapi12experimental6printf"))
+        emitError(WrapperFunc, CI, BadWrapperErrorMsg);
```

Comment:
I guess we can make an early exit from here and not scan other users of a wrapper function